src/Pact/Native.hs

```diff
@@ -1587,6 +1589,26 @@ poseidonHackAChainDef = defGasRNative
         return $ toTerm $ poseidon intArgs
      | otherwise = argsError i as
 
+keccak256Def :: NativeDef
+keccak256Def = defGasRNative
+  "keccak256"
+  keccak256'
+  (funType tTyString [("chunks", TyList tTyString)])
+  [
+    "(keccak256 [\"\"])"
+  , "(keccak256 [\"IP9FQ2ml0FuBp489sFgZ/qmwjCOE91ywq2qhFd1pDaMTGHShyo9witFRnqlSweJJy1QNGWOSx56HdVQk/ufIkICMViciNZ7qUuihL7u5ad15YdK6UgN0k3VaX6BPDVChqibJtEFIwNO5TRxKWaMayhWui9RKy3gz2OkcS4b6MTWkIzh7gVG0Ez7SP21xh7UOwiBK2QGtdNOW5EJ04OyvquF7O5CF4iJgs1ylOxXMUqu6dYr2eY+9BOzuztZI869P2z3tdVeppc+3OCYSqKjz9FlH0aKc4pByko7Bk8ol1RBxvV4ZhOz0AvMG6nYvDyUoL1KW2Zdli+P5g2lv+m0JXGNptNr3nppdMTYikSj462PBK56fp4r/ej6eGaYgIkk80Tbe+7W7e6G5OPNn/S9j61ynbAsP8hueNsPwcjDPPDB05dpYcECnaXXX459ElKzlSG/L84CrdVjE/ollYzW4Lk24ZZUJ6rRqGWExJuWUBCcy3UxBH0GqjN6sccD7QKlObaVYwF53thgoBvJtmv3z2gDGlBkiLIGGpu+tYAtBDmzi8qeX5J3B8TUxmAH6bzlrBvl14qGQoCPkdLYY5w==\"])"
+  , "(keccak256 [\"IP9FQ2ml0FuBp489sFgZ/qmwjCOE91ywq2qhFd1pDaM=\", \"Exh0ocqPcIrRUZ6pUsHiSctUDRljkseeh3VUJP7nyJA=\", \"gIxWJyI1nupS6KEvu7lp3Xlh0rpSA3STdVpfoE8NUKE=\", \"qibJtEFIwNO5TRxKWaMayhWui9RKy3gz2OkcS4b6MTU=\", \"pCM4e4FRtBM+0j9tcYe1DsIgStkBrXTTluRCdODsr6o=\", \"4Xs7kIXiImCzXKU7FcxSq7p1ivZ5j70E7O7O1kjzr08=\", \"2z3tdVeppc+3OCYSqKjz9FlH0aKc4pByko7Bk8ol1RA=\", \"cb1eGYTs9ALzBup2Lw8lKC9SltmXZYvj+YNpb/ptCVw=\", \"Y2m02veeml0xNiKRKPjrY8Ernp+niv96Pp4ZpiAiSTw=\", \"0Tbe+7W7e6G5OPNn/S9j61ynbAsP8hueNsPwcjDPPDA=\", \"dOXaWHBAp2l11+OfRJSs5Uhvy/OAq3VYxP6JZWM1uC4=\", \"TbhllQnqtGoZYTEm5ZQEJzLdTEEfQaqM3qxxwPtAqU4=\", \"baVYwF53thgoBvJtmv3z2gDGlBkiLIGGpu+tYAtBDmw=\", \"4vKnl+SdwfE1MZgB+m85awb5deKhkKAj5HS2GOc=\"])"
+  ]
+  "Compute the hash of a list of base64-encoded inputs. The hash is computed incrementally over all of the base64-decoded inputs."
+  where
+    keccak256' :: RNativeFun e
+    keccak256' i = \case
+      [TList ls _ _] -> do
+        let numBytes = sum $ V.map (\s -> BS.length (T.encodeUtf8 (s ^. _TLitString))) ls
```

Comment:
Should this be gassed before (linear) `encodeUtf8`? I think it's safe to assume ASCII encoding and thus `BS.length ≡ T.length`, since any non-ASCII (and, more specifically, non-base64-set) character will result in a failure anyway.

Reply:
No, we need to pay the cost on the number of base64url-decoded bytes, not the encoded bytes, because we are decoding and then hashing all of the decoded bytes.
That does mean that this calculation is wrong, just in a different way.

Reply:
Actually, I have changed my mind, I think the calculation is correct as-is. I will leave a comment.

Additionally, we don't know that the input is valid base64, so we have to charge for the bytes that are there.

Reply:
Oh, I will also add that, since Text is now UTF-8, `encodeUtf8` is just a memcpy. The difference in their length calculations is that `Text.length` returns length in codepoints, whereas `ByteString.length` returns length in bytes.